src/games.py

```diff
@@ -59,35 +85,115 @@ async def _wait_until_next_announcement():
 
         await asyncio.sleep(next_announcement.total_seconds())
 
+    def _add_epic_games(self):
+        games_to_add = self._get_epic_games()
+        existing_games = db.get_games()
+
+        for game_to_add in games_to_add:
+            name, url = game_to_add["name"], game_to_add["url"]
+            if not contains_substring(url, existing_games):
+                # eg_print(f"{name}: added to next announcement")
+                db.add_game(f"{name}: <https://store.epicgames.com/en-US/p/{url}>")
+
+    @staticmethod
+    def _get_epic_games():
+        """
+        This retrieves epic games newly available for free today. US region only.
+        The parsing is *very* hacky, but to be fair, so is the data format.
+        :return: A list of newly free games.
+        """
+        try:
+            resp = requests.get("https://store-site-backend-static.ak.epicgames.com/freeGamesPromotions?country=US")
+        except Exception as e:
+            eg_print(f"Couldn't get data: {e}")
+            return []
+
+        if resp.status_code != 200:
+            eg_print(f"Couldn't get data: {resp.status_code}: {resp.text}")
+            return []
+
+        try:
+            games = resp.json()["data"]["Catalog"]["searchStore"]["elements"]
+        except Exception as e:
+            eg_print(f"Couldn't parse data: {e}")
+            return []
+
+        # eg_print(f"EG API returned {len(games)} games")
+        games_to_add = []
+        for game in games:
+            try:
+                name = game["title"]
+                url = game["productSlug"]
+                if url is None and len(game["catalogNs"]["mappings"]) > 0:
+                    url = game["catalogNs"]["mappings"][0]["pageSlug"]
+
+                if url is None:
+                    eg_print(f"{name}: Could not get URL, skipping")
+                    continue
+
+                if game["price"]["totalPrice"]["discountPrice"] != 0:
+                    # eg_print(f"{name}: not free, skipping")
+                    continue
+
+                # When there is no promotion the promotions field is a mess:
+                # sometimes null, sometimes an empty object, sometimes a list of no objects
+                # Wrap in try/catch and assume failing to parse it means it's not on sale rather than a parse error
+                try:
+                    start_date = datetime.strptime(
+                        game["promotions"]["promotionalOffers"][0]["promotionalOffers"][0]["startDate"],
+                        "%Y-%m-%dT%H:%M:%S.%fZ")
+                except Exception:
+                    # eg_print(f"{name}: not on sale this week, skipping")
+                    continue
+            except Exception as e:
+                eg_print(f"Couldn't parse game: {e}")
+                continue
+
+            if datetime.utcnow().date() != start_date.date():
```

Comment:
I think Epic games tend to be free for a week or so. Will this repost the same game each day?

Reply:
Oh I didn't read this well. Looks like it'll only post at the first day

Reply:
Yeah, the whole week it retrieves the same games, but only posts them on the day they become available. Added a comment for clarity.